go/vt/automation/scheduler.go

```diff
@@ -0,0 +1,165 @@
+
+// NewScheduler creates a new instance.
+func NewScheduler() (*Scheduler, error) {
+	defaultClusterOperations := map[string]pb.TaskContainerSpec{
+		"reshard": pb.TaskContainerSpec{
+			Type: pb.TaskContainerSpecType_SERIAL_TASKS,
+			SerialTask: []*pb.TaskContainerSpec{
+				&pb.TaskContainerSpec{
+					// TODO(mberlin): Replace this with the actual operations.
+					Type: pb.TaskContainerSpecType_SINGLE_TASK,
+					SingleTask: &pb.TaskSpec{
+						Name: "Shell",
+						Parameter: []*pb.TaskParameter{
+							&pb.TaskParameter{
+								Name:  "command",
+								Value: &pb.Value{Value: []string{"uptime"}},
+							},
+						},
+						OutputKey: "uptime",
+					},
+				},
+			},
+		},
+	}
+
+	return &Scheduler{
+		clusterOperations:              defaultClusterOperations,
+		toBeScheduledClusterOperations: make(chan pb.TaskContainerSpec, 10),
+		globalValues:                   NewValueMap(),
+	}, nil
+}
+
+func (s *Scheduler) registerClusterOperation(name string, op pb.TaskContainerSpec) {
+	// TODO(mberlin): Do not overwrite an existing operation.
+	s.clusterOperations[name] = op
+}
+
+// Run takes care of executing tasks. Blocks until the scheduler is shutdown.
+func (s *Scheduler) Run() {
+	// TODO(mberlin): Off-load executions to a pool of worker.
+	for op := range s.toBeScheduledClusterOperations {
+		s.processTaskContainer(op)
+	}
+}
+
+func (s *Scheduler) processTaskContainer(taskContainer pb.TaskContainerSpec) {
+	// TODO(mberlin): Add missing cases e.g. foreach.
+	switch taskContainer.Type {
+	case pb.TaskContainerSpecType_SINGLE_TASK:
+		s.executeSingleTask(taskContainer.SingleTask)
+	case pb.TaskContainerSpecType_SERIAL_TASKS:
+		for _, serialTaskContainer := range taskContainer.SerialTask {
+			s.processTaskContainer(*serialTaskContainer)
+		}
+	default:
+		// TODO(mberlin): Add error handling.
+	}
+}
+
+func (s *Scheduler) executeSingleTask(singleTask *pb.TaskSpec) error {
+	// TODO(mberlin): Remove debug output.
+	fmt.Printf("executing singleTask: %v\n", singleTask)
+	parameters := make(map[string]*pb.Value)
+
+	for _, taskParameter := range singleTask.Parameter {
+		// TODO(mberlin): Assert that Value is not nil.
+		// TODO(mberlin): Assert that all variables were expanded.
+		input := expandVariables(*taskParameter.Value, s.globalValues)
+		parameters[taskParameter.Name] = &input
+	}
+	singleTask.Input = parameters
+
+	// TODO(mberlin): Check that the list of the task's required parameters is included in the parameters map.
+
+	// TODO(mberlin): Remove debug output.
+	fmt.Printf("running: %v params: %v\n", singleTask.Name, parameters)
+	task := s.createTaskInstance(singleTask.Name)
+
+	output, err := task.run(parameters)
+	if err != nil {
+		return err
+	}
+
+	// TODO(mberlin): Remove debug output.
+	fmt.Printf("output: %v\n", output)
+	singleTask.OutputValue = output
+	s.globalValues.RecordValue(singleTask.OutputKey, *singleTask.OutputValue)
+
+	return nil
+}
+
+func expandVariables(parameter pb.Value, vm *ValueMap) pb.Value {
```

Comment:
This feels fragile, and seems like it could be a source of future pain...

How about using printf format, and letting fmt.Sprintf handling all the details for us? So task value would be something like `"/bin/echo" %v`. And the TaskParameter proto could also contain a repeated string of InputKeys:

&pb.TaskSpec{
  Name: "ShellTask",
  Paramater: []*pb.TaskParameter{
    &pb.TaskParamater{
      Name: "command",
      Value: &pb.Value{Value: []string{"/bin/echo %v"},
      InputKeys: []string{"first-command"},
   }
}

and expandVariables would be simplified to not have to do any string parsing.